### PR TITLE
[https://nvbugs/6442073][fix] Enforce a fixed max_seq_len for Qwen's AttentionOp caching purposes

### PR DESCRIPTION
**Pull request by moraxu**
## Summary by CodeRabbit

- **New Features**
  - Added stable attention capacity handling for multimodal vision encoders.
  - Improved Qwen2.5-VL and Qwen3-VL support across varying image resolutions.
  - Automatically derives encoder capacity from multimodal input token limits when available.

- **Bug Fixes**
  - Added validation for invalid or oversized attention sequences.
  - Improved fallback behavior when multimodal token limits are unavailable.

- **Tests**
  - Added coverage for fixed sequence lengths, window attention limits, capacity calculation, and fallback handling.

## Description

1. The accuracy test creates a long-lived engine and processes 900 MMMU examples over many batches.
2. The test configures an aggregate LLM/decoder engine token budget of 16,384 and an external request batch limit of
   32. The vision encoder inherits 16,384 as nominal metadata sizing, but does not currently enforce it as an admission
   limit.
3. Qwen vision turns each image or temporal video slice into an independent patch-token attention segment.
4. To support potentially many segments, Qwen raises the native request/sequence-slot capacity to 16,384. This provides
   very large practical headroom and explains the semaphore allocation; it is neither the count of active users nor a
   formally proven bound while the encoder token limit remains unenforced.
5. **[PRE-FIX BEHAVIOR]** Every encoder forward assigned the longest **actual** segment in that batch to
   `metadata.max_seq_len`.
6. Different image geometry could yield a previously unseen value of `max_seq_len`.
7. That value was part of the process-static native cache key, together with `layer_idx`.
8. A new value therefore caused up to 27 cache misses, one for each vision layer.
9. Preparing every missed operator allocated a roughly 1 MiB `int32` semaphore array under these capacities.
10. The map had no eviction path, so old operators and their owned arrays remained resident even after their request
    finished.
11. Memory usage therefore grew with the number of distinct batch maxima, not simply with current batch size.
12. Eventually another `reserveSemaphoreArray()` call could no longer allocate device memory, producing the observed
    OOM.

The dense and MoE models use the same 27-layer, 16-head Qwen vision tower, so both are exposed to this mechanism. Their text decoders leave different amounts of runtime memory headroom, which can make them fail at different points.

**[BRANCH FIX]** The fixed Qwen vision sequence-length solution applies to both models. It calculates the
processor-aware bound during model setup, validates each actual segment maximum against it, and assigns the stable
bound to metadata on every forward. The first pass creates one operator per vision layer; later resolutions reuse those
same entries.

**[PRE-EXISTING, PRESERVED BY THE BRANCH FIX]** Exact `seq_lens` and cumulative offsets remain dynamic, so attention correctness is unchanged.

A similar workaround for OOMs is implemented in RADIO: https://github.com/NVIDIA/TensorRT-LLM/blob/0d10ebe25b34509fcc70674fe49cca6261dd00b9/tensorrt_llm/_torch/models/modeling_radio.py#L746-L752

### Diagram 1 - Qwen Vision metadata capacities:

qwen_vision_metadata_capacities

### Diagram 2 - Qwen `AttentionOp` cache growth:

qwen_attention_op_cache_growth

## Test Coverage

```
pytest -sv tests/integration/defs/accuracy/test_llm_api_pytorch_multimodal.py::TestQwen3_5_35B_A3B_VL
pytest -sv tests/integration/defs/accuracy/test_llm_api_pytorch_multimodal.py::TestQwen3_5_27B_VL
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_multimodal_encoder.py [hunk 1]
```diff
@@ -64,6 +64,14 @@ def setup_attn_metadata(self, max_num_requests: int,
             kv_cache_manager=None,
         )
 
+    def set_attn_max_seq_len(self, max_seq_len: int) -> None:
+        """Set an optional stable per-segment attention capacity.
+
+        Specialized encoders may override this hook. Keeping it separate from
```
Comment (moraxu) [inline tensorrt_llm/_torch/models/modeling_multimodal_encoder.py:70]: Not sure if we have a lot of out-of-the-tree encoders so I can move it to `setup_attn_metadata` rather than defining a new method

**Comment by Wanli-Jiang** [comment]
need to unwaive the related tests.

**Review by StanleySun639 — APPROVED** [review]

**Review by chuangz0 — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_multimodal_encoder.py [hunk 2]
```diff
@@ -64,6 +64,14 @@ def setup_attn_metadata(self, max_num_requests: int,
...
+        ``setup_attn_metadata`` preserves that method's existing override
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_multimodal_encoder.py]: Please switch to single backticks on changed lines.

**Review by bo-nv — APPROVED** [review]

File: tests/unittest/_torch/executor/test_pytorch_model_engine.py [hunk 3]
```diff
@@ -188,6 +191,73 @@ def test_spec_decode_capture_restores_kv_lens_between_warmups(self) -> None:
...
+                model_engine.model = torch.nn.Sequential(encoder)
+                model_engine.encoder_batch_size = 32
+                model_engine.encoder_max_num_tokens = 16384
+                model_engine.input_processor = Mock(
+                    spec=BaseMultimodalDummyInputsBuilder)
+                model_engine.input_processor.get_mm_max_tokens_per_item.return_value = demand
+
+                model_engine._set_up_multimodal_encoder_attn_metadata()
+
+                self.assertEqual(encoder.setup_args, (32, 16384))
```
Comment (2ez4bz) [inline tests/unittest/_torch/executor/test_pytorch_model_engine.py]: Nit: can these values be captured as args and used across lines 221 and 229?

Same comment for the other test function.

File: tests/unittest/_torch/executor/test_pytorch_model_engine.py [hunk 4]
```diff
@@ -188,6 +191,73 @@ def test_spec_decode_capture_restores_kv_lens_between_warmups(self) -> None:
...
+                self.assertEqual(encoder.max_seq_len, expected_max_seq_len)
+
+    def test_multimodal_encoder_max_seq_len_falls_back_without_dummy_builder(
```
Comment (2ez4bz) [inline tests/unittest/_torch/executor/test_pytorch_model_engine.py]: Could this be folded into the previous test function via parameterization on the inputs + expected values?

File: tests/unittest/_torch/modeling/test_modeling_qwen2_5vl.py [hunk 5]
```diff
@@ -496,6 +497,96 @@ def test_prepare_qwen_vl_mrope_config_pure_generation_reads_deltas_only():
...
+    with pytest.raises(ValueError, match=r"exceeds.*maximum=65536"):
+        _prepare_qwen_vl_vision_attn_metadata([128, 65537],
+                                              metadata,
+                                              max_seq_len=65536)
+
+    assert metadata.prepare_encoder_only_calls == 0
+
+
+def test_qwen3_vision_prepare_metadata_passes_fixed_max_seq_len() -> None:
+    model = Qwen3VisionModel.__new__(Qwen3VisionModel)
```
Comment (2ez4bz) [inline tests/unittest/_torch/modeling/test_modeling_qwen2_5vl.py]: Should this and under qwen3 VL tests be in `test_modeling_qwen3vl.py` instead?

File: tests/unittest/_torch/modeling/test_modeling_qwen2_5vl.py [hunk 6]
```diff
@@ -496,6 +497,96 @@ def test_prepare_qwen_vl_mrope_config_pure_generation_reads_deltas_only():
...
+    metadata = _StubVisionAttentionMetadata()
+
+    model.prepare_attn_metadata([256, 1024], metadata)
+
+    assert metadata.max_seq_len == 65536
+    assert metadata.seq_lens.tolist() == [256, 1024]
+
+
+def test_qwen2_5_window_attention_uses_tighter_fixed_max_seq_len() -> None:
+    model = Qwen2_5_VisionModel.__new__(Qwen2_5_VisionModel)
```
Comment (2ez4bz) [inline tests/unittest/_torch/modeling/test_modeling_qwen2_5vl.py]: Can we instantiate the actual object instead of mocking? We have a lightweight config in this file that can be used.

I suspect lines 575-578 would not be necessary anymore as a result.

File: tests/unittest/_torch/modeling/test_modeling_qwen2_5vl.py [hunk 7]
```diff
@@ -496,6 +497,96 @@ def test_prepare_qwen_vl_mrope_config_pure_generation_reads_deltas_only():
...
+    model.spatial_merge_unit = 4
+    model.patch_size = 14
+    model.metadata_cls = _StubVisionAttentionMetadata
+    model.patch_embed = SimpleNamespace(proj=SimpleNamespace(
+        weight=torch.empty(1)))
+
+    model.setup_attn_metadata(max_num_requests=8, max_num_tokens=16)
+    model.set_attn_max_seq_len(65536)
+
+    assert model._full_attn_max_seq_len == 65536
```
Comment (2ez4bz) [inline tests/unittest/_torch/modeling/test_modeling_qwen2_5vl.py]: Nit: same comment regarding the usage of variables for lines 584 and 586.

**Review by 2ez4bz — APPROVED** [review]
Do I understand it correctly that the mechanism that fixes the OOM is using a fixed `max_seq_len` for the attention op?

If so, does that mean using the LLM's `max_seq_len` would solve that OOM issue? But, we might as well use the value specified by the multimodal encoder specific args?

**Comment by moraxu** [comment]
> Do I understand it correctly that the mechanism that fixes the OOM is using a fixed `max_seq_len` for the attention op?
> 
> If so, does that mean using the LLM's `max_seq_len` would solve that OOM issue? But, we might as well use the value specified by the multimodal encoder specific args?

Yes to the first question, but for VLMs the relevant arg is `encoder_max_num_tokens` (nut sure if that's the one you meant) which is currently not enforced (I believe that's what Yechan's PR implements: https://github.com/NVIDIA/TensorRT-LLM/pull/16051), and Qwen's `max_seq_len` variable doesn't use LLM’s `max_seq_len`.

**Review by LarryXFly — APPROVED** [review]

**Review by zhaoyangwang-nvidia — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]
